annnufan/assignment2/node.cpp

```diff
@@ -0,0 +1,96 @@
+#include <iostream>
+#include <algorithm>
+
+
+class node {
+	int key;
+public:
+	node(int x, node* l = nullptr, node* r = nullptr) : key(x), left_node(l), right_node(r){}
+	~node() {
+		delete left_node;
+		delete right_node;
+	}
+
+	void set_value(int x) {
+		key = x;
+	}
+
+	int get_value() {
+		return key;
+	}
+
+	node* left_node, *right_node, *parent_node;
+};
+
+bool find_key_and_print_way_to_root(int x, node* v) {
+	if (v == nullptr) {
+		return false;
+	}
+	if (x == v->get_value()) {
+		return true;
+	}
+	if (find_key_and_print_way_to_root(x, v->left_node) || find_key_and_print_way_to_root(x, v->right_node)) {
+		std::cout << v->get_value() << ' ';
+		return true;
+	}
+	return false;
+}
+
+void print_way_to_root_for_key(int x, node* v) {
+	if (find_key_and_print_way_to_root(x, v)) {
+		std::cout << "\n";
+	} else {
+		std::cout << "Couldn't find key\n";
+	}
+}
+
+node* find_node(int x, node* v, node* p) {

```

Comment:
This function not only finds nodes, but also fills the parent data.
From the perspective of a user, this can be unexpected; The user expects this function to return the node which has the key `x`. But this function also changes the state of the tree!
Also with the current function signature, it is not clear what the parameter `p` does.
A better name would be `find_node_and_update_parent`.

But even better:
Code is most readable and easy to reuse when each function does only one thing.
So you can split it to two functions:
`find_node` and `update_parents`, each one does just the one thing that you expect it to do from its name.

-----
But let's take a look at the `parent_node` field. When is it set?
Because it is sometimes set and sometimes not, it is hard for the user to follow when it is set and when they can use the functions that depend on it (e.g. `count_high`).
So I suggest that you update the `parent_node` field in the constructor; this way the user does not need to think - did I update this field? Did I forget? Why the code does not work? Because I forgot to update the parents! etc...

The constructor can work like this:
```C++
node(int x, node* l, node* r) : key(x), left(l), right(r) {
    if (left != null) left->parent = this;
    if (right != null) right->parent = this;
}

node* root = new node(1, new node(2, nullptr, nullptr), new node(3, nullptr, nullptr));
// After the line above, nodes 2 and 3 already have their parent field set correctly.
```
`parent` can stay private field because you only modify it inside `node` constructor.